src/core/ddsc/tests/deadline.c

```diff
@@ -434,3 +437,125 @@ CU_Theory((int32_t n_inst, uint8_t unreg_nth, uint8_t dispose_nth), ddsc_deadlin
+
+  dds_return_t rc = dds_set_status_mask(wr, DDS_PUBLICATION_MATCHED_STATUS);
+  CU_ASSERT_EQUAL_FATAL(rc, DDS_RETCODE_OK);
+
+  uint32_t status = 0;
+  while(!(status & DDS_PUBLICATION_MATCHED_STATUS))
+  {
+    /* Polling sleep. */
+    dds_sleepfor (DDS_MSECS(1));
+
+    CU_ASSERT_EQUAL_FATAL(rc, DDS_RETCODE_OK);
+    rc = dds_get_status_changes (wr, &status);
+  }
+
+  struct ddsi_domaingv *gvptr = get_domaingv (wr);
+  ddsrt_mtime_t now = ddsrt_time_monotonic();
+  ddsrt_mtime_t next = ddsrt_mtime_add_duration(now, DEADLINE);
+  struct ddsi_xevent *xev = ddsi_qxev_callback(gvptr->xevents, next, cb, NULL);
+  CU_ASSERT_FATAL(xev != NULL);
+
+  /*write*/
+  Space_Type1 msg1 = { 1, 0, 0 }, //this is the main instance whose deadline will expire but get pushed back
+              msg2 = { 2, 0, 0 }; //this is a secondary instance whose deadline will get pushed back all the time
+  rc = dds_write(wr, &msg1);  /* expires @ DEADLINE */
+  CU_ASSERT_EQUAL_FATAL(rc, DDS_RETCODE_OK);
+  rc = dds_write(wr, &msg2);  /* expires @ DEADLINE */
+  CU_ASSERT_EQUAL_FATAL(rc, DDS_RETCODE_OK);
+
+  dds_instance_handle_t ih = dds_lookup_instance(wr, &msg1);
+
+  dds_requested_deadline_missed_status_t rstatus;
+
+  rc = dds_get_requested_deadline_missed_status (rd, &rstatus);
+  CU_ASSERT_EQUAL_FATAL(rc, DDS_RETCODE_OK);
+  CU_ASSERT_EQUAL (rstatus.total_count, 0);
+  CU_ASSERT_EQUAL (rstatus.last_instance_handle, 0);
+
+  dds_offered_deadline_missed_status_t ostatus;
+  rc = dds_get_offered_deadline_missed_status (wr, &ostatus);
+  CU_ASSERT_EQUAL_FATAL(rc, DDS_RETCODE_OK);
+  CU_ASSERT_EQUAL (ostatus.total_count, 0);
+  CU_ASSERT_EQUAL (ostatus.last_instance_handle, 0);
+
+  dds_sleepfor(DEADLINE/2);
+  rc = dds_write(wr, &msg2);  /* expires @ 1.5*DEADLINE */
+  CU_ASSERT_EQUAL_FATAL(rc, DDS_RETCODE_OK);
+  dds_sleepfor(DEADLINE/2);
+
+  rc = dds_write(wr, &msg1);  /* expires @ 2*DEADLINE */
+  CU_ASSERT_EQUAL_FATAL(rc, DDS_RETCODE_OK);
+  rc = dds_write(wr, &msg2);  /* expires @ 2*DEADLINE */
+  CU_ASSERT_EQUAL_FATAL(rc, DDS_RETCODE_OK);
+
+  dds_sleepfor(DEADLINE/2);
+  rc = dds_write(wr, &msg2);  /* expires @ 2.5*DEADLINE */
+  CU_ASSERT_EQUAL_FATAL(rc, DDS_RETCODE_OK);
+
+  rc = dds_get_requested_deadline_missed_status (rd, &rstatus);
+  CU_ASSERT_EQUAL_FATAL(rc, DDS_RETCODE_OK);
+/* here the status can be updated already, or queued to be updated
+   on the expiry of the next deadline */
+  CU_ASSERT ((rstatus.total_count == 0 && rstatus.last_instance_handle == 0) ||
+             (rstatus.total_count == 1 && rstatus.last_instance_handle == ih));
+
+  rc = dds_get_offered_deadline_missed_status (wr, &ostatus);
+  CU_ASSERT_EQUAL_FATAL(rc, DDS_RETCODE_OK);
+/* here the status can be updated already, or queued to be updated
+   on the expiry of the next deadline */
+  CU_ASSERT ((ostatus.total_count == 0 && ostatus.last_instance_handle == 0) ||
+             (ostatus.total_count == 1 && ostatus.last_instance_handle == ih));
+
+  dds_sleepfor(DEADLINE/2);
```

Comment:
Handler runs (if I did my math correct)! Notifies the reader because T happens to be at the head of the list, updates the list to [U,T]